django/test/client.py

```diff
@@ -320,8 +320,12 @@ def _encode_json(self, data, content_type):
         Return encoded JSON if data is a dict and content_type is
         application/json.
         """
-        should_encode = JSON_CONTENT_TYPE_RE.match(content_type) and isinstance(data, dict)
-        return json.dumps(data, cls=self.json_encoder) if should_encode else data
+        if JSON_CONTENT_TYPE_RE.match(content_type) and not isinstance(data, str):
+            try:
+                return json.dumps(data, cls=self.json_encoder)
+            except TypeError:
+                pass
```

Comment:
I feel like this should raise an exception and not be swallowed silently:

```
In [1]: class Abc:
   ...:     pass

In [2]: import json

In [3]: json.dumps([Abc()])
...
TypeError: Object of type Abc is not JSON serializable
```

If `[Abc()]` is passed here it would result in `[Abc()]` being returned, which might error somewhere down the line in a weird fashion, or worse end up with `str([Abc()])` being sent.

Reply:
I thought about this, and happy to go with that, but I thought that falling back to the rest of the data handling logic applying as normal could be also considered a good thing.

For example, currently using `data=json.dumps({}).encode('utf-8')` would work, but if this threw the `TypeError` it would fail, certainly breaking backwards compatibility with our tests, and possibly with other use-cases.

I'm planning to leave as-is for this reason, but do let me know if you feel strongly enough! If so I'd appreciate advice on how this should be documented.

Reply:
> currently using `data=json.dumps({}).encode('utf-8')` would work

Yes but `bytes` as body is supported as the default for other content types, and `bytes` aren't JSON serializable either. I think if you ignored both `str` and `bytes` on 323 then we could probably raise the exceptions fine?

> certainly breaking backwards compatibility with our tests

"our" = Django's or Thread's? Have you got any other examples?


Alternatively you could be changing the logic on line 323 to increase the whitelist of allowed types - i.e. add `list` and perhaps `int` / `float` / `bool` - rather than changing it to a blacklist of types *not* to `json.dumps`. Granted there may be edge cases for other json encoders, like `Decimal`, but it's safer for everyone's tests, and more obvious than swallowing an exception.

Reply:
I feel similarly. Maybe it's better just to add support for list and tuple as originally proposed. It's unclear to me if other types would actually be used.

Reply:
Only `list` and `tuple` only would be fine by me, you're right it's unclear the other default json types would be used.

Reply:
Great. I don't feel strongly. I'll roll this back to just list/tuple/dict and update.